File: 백트래킹/실버/실버2/6603_로또/6603.py
import sys

def main(queries):
    for idx, query in enumerate(queries):
        if query[0] == 0:
            return
        lotto_list = query[1:]
        print_combinations(lotto_list)
        backtrack(6, [], lotto_list, 0)
        if idx < len(queries) - 2:
            print()

def print_combinations(lotto_list):
    sequence_length = 6
    backtrack(sequence_length, [], lotto_list, 0)

def backtrack(sequence_length, sequence, lotto_list, start):
    if len(sequence) == sequence_length:
        print(" ".join(map(str, sequence)))
        return

    for i in range(start, len(lotto_list)):
        sequence.append(lotto_list[i])
        backtrack(N, sequence, lotto_list, i + 1)
        sequence.pop()
    
def read_input():
    input_data = sys.stdin.read()
    lines = input_data.splitlines()
    queries = [list(map(int, line.split())) for line in lines]

if __name__ == "__main__":
    queries = read_intput()
    main(queries)

File: 백트래킹/실버/실버2/6603_로또/6603.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        List<List<Integer>> queries = readInput(sc);

        for (int idx = 0; idx < queries.size(); idx++) {
            List<Integer> query = queries.get(idx);
            if (query.get(0) == 0) {
                return;
            }

            List<Integer> lottoList = query.subList(1, query.size());
            printCombinations(lottoList);

            if (idx < queries.size() - 2) {
                System.out.println();
            }
        }

        sc.close();
    }

    public static void printCombinations(List<Integer> lottoList) {
        int sequenceLength = 6;
        backtrack(sequenceLength, new ArrayList<>(), lottoList, 0);
    }

    public static void backtrack(int sequenceLength, List<Integer> sequence, List<Integer> lottoList, int start) {
        if (sequence.size() == sequenceLength) {
            System.out.println(String.join(" ", sequence.toString().replaceAll("[\\[\\],]", "")));
            return;
        }

        for (int i = start; i < lottoList.size(); i++) {
            sequence.add(lottoList.get(i));
            backtrack(sequenceLength, sequence, lottoList, i + 1);
            sequence.remove(sequence.size() - 1);
        }
    }

    public static List<List<Integer>> readInput(Scanner sc) {
        List<List<Integer>> queries = new ArrayList<>();
        while (sc.hasNextLine()) {
            String line = sc.nextLine();
            if (line.trim().isEmpty()) {
                break;
            }

            String[] tokens = line.split(" ");
            List<Integer> query = new ArrayList<>();
            for (String token : tokens) {
                query.add(Integer.parseInt(token));
            }
            queries.add(query);
        }
        return queries;
    }
}
